src/ripple/app/misc/NegativeUNLVote.h

```diff
@@ -0,0 +1,189 @@
+     * Cast our local vote on the NegativeUNL candidates.
+     *
+     * @param prevLedger the parent ledger
+     * @param unlKeys the trusted master keys
+     * @param validations the validation message container
+     * @param initialSet the set of transactions
+     */
+    void
+    doVoting(
+        LedgerConstPtr& prevLedger,
+        hash_set<PublicKey> const& unlKeys,
+        RCLValidations& validations,
+        std::shared_ptr<SHAMap> const& initialSet);
+
+    /**
+     * Notify NegativeUNLVote that new validators are added.
+     * So that they don't get voted to the NegativeUNL immediately.
+     *
+     * @param seq the current LedgerIndex when adding the new validators
+     * @param nowTrusted the new validators
+     */
+    void
+    newValidators(LedgerIndex seq, hash_set<NodeID> const& nowTrusted);
+
+private:
+    NodeID const myId_;
+    beast::Journal j_;
+    std::mutex mutex_;
+    hash_map<NodeID, LedgerIndex> newValidators_;
+
+    /**
+     * Add a ttUNL_MODIFY Tx to the transaction set.
+     *
+     * @param seq the LedgerIndex when adding the Tx
+     * @param vp the master public key of the validator
+     * @param disabling disabling or re-enabling the validator
+     * @param initialSet the transaction set
+     */
+    void
+    addTx(
+        LedgerIndex seq,
+        PublicKey const& vp,
+        bool disabling,
+        std::shared_ptr<SHAMap> const& initialSet);
+
+    /**
+     * Pick one candidate from a vector of candidates.
+     *
+     * @param randomPadData the data used for picking a candidate.
+     *        @note Nodes must use the same randomPadData for picking the same
+     *        candidate. The hash of the parent ledger is a good choice.
+     * @param candidates the vector of candidates
+     * @return the picked candidate
+     */
+    NodeID
+    pickOneCandidate(uint256 randomPadData, std::vector<NodeID>& candidates);
+
+    /**
+     * Build a reliability measurement score table of validators' validation
+     * messages in the last flag ledger period.
+     *
+     * @param prevLedger the parent ledger
+     * @param unl the trusted master keys
+     * @param validations the validation container
+     * @param scoreTable the score table to be built
+     * @return if the score table was successfully filled
+     */
+    bool
+    buildScoreTable(
+        LedgerConstPtr& prevLedger,
+        hash_set<NodeID> const& unl,
+        RCLValidations& validations,
+        hash_map<NodeID, unsigned int>& scoreTable);
+
+    /**
+     * Process the score table and find all disabling and re-enabling
+     * candidates.
+     *
+     * @param unl the trusted master keys
+     * @param nUnl the NegativeUNL
+     * @param scoreTable the score table
+     * @param toDisableCandidates the candidates to disable
+     * @param toReEnableCandidates the candidates to re-enable
+     */
+    void
+    findAllCandidates(
+        hash_set<NodeID> const& unl,
+        hash_set<NodeID> const& nUnl,
```

Comment:
People sometimes use an `n` prefix to denote `num`. Maybe rename this `negUnl`?

Reply:
Thanks, changed. Now I realize there are other `nUnl*` in the code. I might change this back or change the others to `negUnl*` too.